vault/resource_pki_backend_role_test.go

```diff
@@ -0,0 +1,236 @@
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
+	})
+}
+
+func TestAccPKIBackendRole(t *testing.T) {
+	backend := acctest.RandomWithPrefix("pki")
+	role := acctest.RandomWithPrefix("test-role")
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testProviders,
+		CheckDestroy: testAccCheckPKIBackendRoleDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccPKIBackendRoleConfig(backend, role),
+				Check:  testAccPKIBackendRoleCheck_attrs(backend, role),
+			},
+		},
+	})
+}
+
+func TestAccPKIBackendRoleUpdate(t *testing.T) {
+	backend := acctest.RandomWithPrefix("pki")
+	role := acctest.RandomWithPrefix("test-role")
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testProviders,
+		CheckDestroy: testAccCheckPKIBackendRoleDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccPKIBackendRoleConfig(backend, role),
+				Check:  testAccPKIBackendRoleCheck_attrs(backend, role),
+			},
+			{
+				Config: testAccPKIBackendRoleConfigUpdate(backend, role),
+				Check: resource.ComposeTestCheckFunc(
+					testAccPKIBackendRoleCheck_attrs(backend, role),
+					resource.TestCheckResourceAttr("vault_pki_backend_role.role",
+						"ttl", "48h0m0s"),
+					resource.TestCheckResourceAttr("vault_pki_backend_role.role",
+						"max_ttl", "72h0m0s"),
+					resource.TestCheckResourceAttr("vault_pki_backend_role.role",
+						"allow_ip_sans", "false"),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckPKIBackendRoleDestroy(s *terraform.State) error {
+	client := testProvider.Meta().(*api.Client)
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "vault_pki_backend_role" {
+			continue
+		}
+		secret, err := client.Logical().Read(rs.Primary.ID)
+		if err != nil {
+			return fmt.Errorf("Error checking for PKI backend role %q: %s", rs.Primary.ID, err)
+		}
+		if secret != nil {
+			return fmt.Errorf("PKI backend role %q still exists", rs.Primary.ID)
+		}
+	}
+	return nil
+}
+
+func testAccPKIBackendRoleCheck_attrs(backend, role string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		resourceState := s.Modules[0].Resources["vault_pki_backend_role.role"]
+		if resourceState == nil {
+			return fmt.Errorf("resource not found in state")
+		}
+
+		instanceState := resourceState.Primary
+		if instanceState == nil {
+			return fmt.Errorf("resource not found in state")
+		}
+
+		endpoint := instanceState.ID
+
+		if endpoint != backend+"/roles/"+role {
+			return fmt.Errorf("expected ID to be %q, got %q instead", backend+"/roles/"+role, endpoint)
+		}
+
+		client := testProvider.Meta().(*api.Client)
+		resp, err := client.Logical().Read(endpoint)
+		if err != nil {
+			return fmt.Errorf("%q doesn't exist", endpoint)
+		}
+
+		attrs := map[string]string{
+			"ttl":     "ttl",
+			"max_ttl": "max_ttl",
+		}
+		for stateAttr, apiAttr := range attrs {
```

Comment:
I'm fairly confident you're copying a bad idea of mine on this. I _do_ regret doing tests this way, and would rather not do them like this in the future. It _is_ a valid way to test though, so I don't want to block this PR on it, I just wanted to call out that I think we should move away from code like this in the future.

Reply:
🤔 Yeah, I figured this was the "tf provider way". 

LMK if you think its important to change, otherwise I think everything else is good to go.